017-storage-buckets.md

```diff
@@ -0,0 +1,240 @@
+- **Adapter**
+  - This decides to what Storage Adapter the files will be uploaded to. Possible options are Local, S3, DigitaOcean Spaces, etc. This is not allowed to be changed after a file has been added to the bucket.
+- **Maximum File Size Limit**
+  - This setting decides the maximum single file size that can be uploaded.
+- **Maximum Number of Files**
+  - This setting decides how many files a Bucket can contain.
+- **Accepted File Extensions**
+  - This setting decides what file extensions can be uploaded.
+- **Encryption**
+  - This setting decides if files are encrypted. This will come into play, when encrypting large files causes too much load. This is not allowed to be changed after a file has been added to the bucket.
+- **Virus Scan**
+  - This setting decides if files are scanned by ClamAV after upload. Same reasoning as **Encryption**.
+- **TTL**
+  - This setting decides how long a file is supposed to alive. Files are going to be deleted automatically. This is not allowed to be changed after a file has been added to the bucket.
+- **Permission**
+  - This setting decides what permission role is allowed to upload files to the bucket. Read and Write permissions will still exist for each file and be the source of permission check.
+
+A Bucket will have following Model in our internal Database:
+```typescript
+{
+  $id: string;
+  $write: string[];
+  name: string;
+  enabled: boolean;
+  adapter: 'local'|'s3'|'etc'
+  encrypted: boolean;
+  antivirus: boolean;
+  ttl?: number;
+  maximumFiles?: number;
+  maximumFileSize?: number;
+  allowedFileExtensions?: string;
+}
+```
+
+In terms of a storage adapter except `Local`, additional values will be present like configurations for S3.
+
+### New Endpoints
+
+In this section I will explain all necessary endpoints for integratting Storge Buckets.
+
+#### `GET: /v1/storage/buckets/`
+
+This endpoint lists all Buckets. *Server*
+
+Payload:
+- **search = ''** Search term to filter your list results.
+- **limit = 25** Maximum number of Buckets to return.
+- **offset = 0** Offset value.
+
+#### `GET: /v1/storage/buckets/{bucketId}`
+
+This endpoint returns a single Bucket by its unique ID. *Server*
+
+#### `GET: /v1/storage/buckets/{bucketId}/files`
+
+This endpoint returns a list of all Files in a specific Bucket. *Server & Client*
+
+Payload:
+- **filters = []** Array of filter strings.
+- **limit = 25** Maximum number of Buckets to return.
+- **offset = 0** Offset value.
+- **search = ''** Search query.
+
+#### `POST: /v1/storage/buckets/{adapter}`
+
+This endpoint creates a new Bucket. *Server*
+
+Payload:
+- **name** Bucket name.
+- **write** An array of strings with write permissions.
+- **encrypted = true** Enable Encryption.
+- **antivirus = true** Enable Anti Virus.
+- **ttl = false** Enables TTL for files.
+- **maximumFiles = null** Maximum amount of files.
+- **maximumFileSize = null** Maximum file size.
+- **allowedFileExtensions = null** Allowed File extensions.
```

Comment:
```suggestion
- **name** Bucket name.
- **write** An array of strings with write permissions.
- **encrypted = true** Enable Encryption.
- **antivirus = true** Enable Anti Virus.
- **ttl = false** Enables TTL for files.
- **maxFiles = null** Maximum amount of files.
- **maxFileSize = null** Maximum file size.
- **fileExtensions = null** Allowed File extensions. Leave empty to allow any extension.
```

Added a few suggestions to param names - still not 100% sure about them.

I think we should consider to implement bot `maxFiles` and `ttl` options in a later stage, not sure how well the DB will scale with them.